### ISSUE TITLE
Expose Return Code as a variable/set a custom variable

### ISSUE THREAD

**Issue by peternewman** [bitfocus/companion-module-generic-http#63]
It would be great to be able to get the return code (e.g. to style a button more obviously if a request failed) rather than just the red square/connection failure. I guess this would need to set another custom variable as multiple requests could be in flight simultaneously.

**Comment by MikeDFWM** [comment]
I second this--the ability to set return codes as variables would allow for triggers based on those return codes, which would be immensely helpful for automating backup plans in case of failures.

**Comment by jswalden** [comment]
Setting another custom variable merely transfers the problem of multiple instances of an action being in flight at the same time, because an action doesn't inherently block other instances of that action from executing.

Realistically, what is needed is some sort of way to _chain_ one action to another, such that the first action can produce some kind of result value that is passed as an input to the second action.  (And I guess the first action in such chain could be another chain, so that you can create actual pipelines out of this.)  And that would have to be a Companion-level feature request/enhancement of existing actions.  I've searched around for an issue that would cover this -- I'm sure there must be one that could encompass it -- but haven't managed to find it myself.

**Comment by MikeDFWM** [comment]
This is 100% an XY Problem.

And, yes, the best way to handle it would probably be an update to Companion itself.

However, in that context, it'd be a fairly niche request that'd be fairly far down on the roadmap if it ever gets done at all.

An option within this module similar to that which exists for GET returns, but for Return Status Codes, to optionally save them as a selected variable, may not work for everyone in every situation, but it would presumably be far simpler & quicker to implement & would be something that many of us can work with, especially while we wait to see if/when Companion's core functionality is updated to provide other means to achieve similar ends.

**Comment by peternewman** [comment]
> Setting another custom variable merely transfers the problem of multiple instances of an action being in flight at the same time, because an action doesn't inherently block other instances of that action from executing.

No, agreed, but for many use cases, its possible to do that externally, e.g. inhibiting the action until a response has occurred, or as was the case in my use case, it's not a repeated press anyway, just a one off. As @MikeDFWM says, it would fix it for those of us who have simpler requirements.

> Realistically, what is needed is some sort of way to _chain_ one action to another, such that the first action can produce some kind of result value that is passed as an input to the second action. (And I guess the first action in such chain could be another chain, so that you can create actual pipelines out of this.) And that would have to be a Companion-level feature request/enhancement of existing actions. I've searched around for an issue that would cover this -- I'm sure there must be one that could encompass it -- but haven't managed to find it myself.

I think this is effectively the issue that would solve it https://github.com/bitfocus/companion/issues/2827 which has now been merged into https://github.com/bitfocus/companion/issues/1909 .

But I'd also argue that if that was implemented, the action would still need the means to set a variable with the return code, it's just the scope of that variable that would change.

**Comment by mw-pages-8621583** [comment]
> But I'd also argue that if that was implemented, the action would still need the means to set a variable with the return code, it's just the scope of that variable that would change.

I was imagining that the idea would be to update Companion to take the value from the log itself. That's why I described it as more universally helpful, but niche.

If it's actually using a literal variable, then, yeah, we just took a long path back to where we already were, because you need to start with being be able to get a variable out of the return status code.

That's just shifting the question to "what will you do with it when you get it", which is going to be decided on a case-by-case basis anyway.
